doc/website/blog/2022-03-31-ent-hooks-examples.md

```diff
@@ -0,0 +1,411 @@
+
+### Keeping our secrets safe:
+
+Many times, we need to store strings that include secrets, a good example can be a phone number.
+In our case the privacy concern is rather simple. We require to hide the last 4 digits of each phone number.
+The expected behavior is best described by this test:
+```go title="ent/schema/user_test.go"
+func TestUserPhoneNumberHook(t *testing.T) {
+	ctx := context.Background()
+	c := enttest.Open(t, dialect.SQLite, "file:TestSchemaConfHooks?mode=memory&cache=shared&_fk=1")
+	u := c.User.Create().SetName("Yoni").SetPhoneNumber("315-194-6020").SaveX(ctx)
+	require.Equal(t, "315-194-****", u.PhoneNumber)
+	require.Equal(t, "6020", u.LastDigits)
+	require.Equal(t, "315-194-6020", u.FullPhoneNumber())
+}
+```
+
+As you can see, inserting a phone number updates the string to have `****` instead of the last digits and stores them 
+in the `last_digits` field that is defined as [sensitive](https://entgo.io/docs/schema-fields#sensitive-fields).
+To have the full phone number we created a utility function `FullPhoneNumber()`.
+
+#### Recipe:
+
+First thing, add the hook to your schema:
+```go title="ent/schema/user.go"
+// Hooks of the User.
+func (User) Hooks() []ent.Hook {
+	return []ent.Hook{
+		hook.If(maskPhoneNumber,
+			hook.And(hook.Or(hook.HasOp(ent.OpUpdateOne), hook.HasOp(ent.OpCreate)), hook.HasFields(user.FieldPhoneNumber)),
+		),
+	}
+}
+```
+* Run `go generate ./ent` or the hook will not be registered.
+
+A few [hook helpers](https://entgo.io/docs/hooks/#hook-helpers) are used here, `maskPhoneNumber` is the name of the hook function, and it will run if a user 
+is updated or created and the mutation has the field `phone_number` (if we change only the name of the user there is no need to do anything right?).
+
+The hook:
+```go title="ent/schema/user.go"
+func maskPhoneNumber(next ent.Mutator) ent.Mutator {
+	return hook.UserFunc(func(ctx context.Context, m *gen.UserMutation) (ent.Value, error) {
+		cs, ok := m.PhoneNumber()
+		if !ok {
+			return next.Mutate(ctx, m)
+		}
+		sp := strings.Split(cs, "-")
+		if len(sp) != 3 {
+			return next.Mutate(ctx, m)
+		}
+		m.SetLastDigits(sp[2])
+		sp[2] = "****"
+		m.SetPhoneNumber(strings.Join(sp[0:3], "-"))
+		return next.Mutate(ctx, m)
+	})
+}
+```
+We validate that the mutation has a `phone_number`(if not just let the mutation continue).
+A bit of string acrobatics, and we have the `last_digits` extracted from the string. We store it in another field - the `last_digits` and
+mask the `phone_number`. 
+
+To get the full phone number we add a utility function 
+```go title="ent/user_phone_number.go"
+package ent
+
+import "strings"
+
+// FullPhoneNumber returns the phone number unmasked.
+func (u *User) FullPhoneNumber() string {
+	return strings.ReplaceAll(u.PhoneNumber, "****", u.LastDigits)
+}
+```
+
+It's valid to add files to the Ent directory (not only in schema folder) since the Ent generate command only appends/changes Ent
```

Comment:
```suggestion
It's valid to add files to the Ent directory (not only in the schema folder) since the Ent `generate` command only appends/changes Ent
```